iOS-TvShows/iOS-TvShows/Common/Services/AppKeyboardService/KeyboardViewController.swift

```diff
@@ -0,0 +1,49 @@
+//
+//  KeyboardViewController.swift
+//  iOS-TvShows
+//
+//  Created by Filip Ban on 10.11.2021..
+//
+
+import UIKit
+import SnapKit
+
+class KeyboardViewController: UIViewController, KeyboardSubscriberInterface {
```

Comment:
While it is cool to have shared logic between view controllers for handling keyboard, but composition is a better approach rather than subclassing. Sometimes you'll have the case where your view controller has a specific subclass and you won't be able to add subclass for your keyboard view controller.

```swift
protocol KeyboardViewControllerSubscriberInterface: KeyboardSubscriberInterface {
    var bottomConstraint: Constraint? { get }
}

extension KeyboardViewControllerSubscriberInterface where Self: UIViewController {

    func onKeyboardFrameChanged(
        endFrame: CGRect,
        endFrameY: CGFloat,
        animationDuration: Double,
        bottomPadding: CGFloat
    ) {
        if endFrameY >= UIScreen.main.bounds.size.height {
            bottomConstraint?.update(offset: 0)
        } else {
            let frameHeight = endFrame.size.height
            bottomConstraint?.update(offset: bottomPadding - frameHeight)
        }

        UIView.animate(withDuration: animationDuration) { [weak self] in
            self?.view.layoutIfNeeded()
        }
    }
}

class SomeViewController: KeyboardViewControllerSubscriberInterface {

    var bottomConstraint: Constraint? { view.someConstraint ... }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(...)
        appKeyboardService?.subscribe(subscriber: self)
    }
}
```

Reply:
Also, what about scrollviews where you don't want to change bottom constraint, rather content inset?